### ISSUE TITLE
Subvolumes with arbitrarily-non-conformant characters in the path result in validation error

### ISSUE THREAD

**Issue by zatricky** [digint/btrbk#120]
**Background**:
Per previous issue #3 , the config file has "validation" that necessitates a workaround if someone wants to use "funky" characters in the path. But it seems to me that the validation is naive. Arbitrary paths should be possible without a whitelist of "valid" characters.

**Scenario**
In this particular case I want a separate snapshot/backup of the "~/VirtualBox VMs" folder which is a subvolume. The result is an error due to the space character in the path:

```
[root@test ~]# /usr/sbin/btrbk -vc /etc/btrbk/test-server.conf dryrun
btrbk command line client, version 0.24.0  (Thu Dec 15 10:51:52 2016)
Using configuration: /etc/btrbk/test-server.conf
ERROR: Ambiguous file for option "subvolume" in "/etc/btrbk/test-server.conf" line 144: "home/user/VirtualBox VMs"
ERROR: Failed to parse configuration file
[root@test ~]#
```

**Proposed solution**
I can do a workaround by renaming the subvol and using a symlink - but that isn't KISS. The right thing to do is to change path validation, which would be cleaner and has better guarantees against bad configs:

- Allow for quoted paths
- Add reasonable conditions, such as requiring a direct match against the canonical sub list*
- Remove the arbitrarily-weird "character" validation on paths.

*Along the lines that the path matches exactly one of the results from "btrfs sub list" without interpretation (i.e. don't interpret "path/../real-path". Simply pattern-match it against "btrfs sub list". The sub list will *never* have "path/../real-path" in the result-list.

**On Security**
While there certainly are security concerns over paths, if a root user makes bad security decisions, that shouldn't be btrbk's responsibility. Long ropes are more useful than short ropes. Don't disallow long ropes for fear of users hanging themselves.

**Comment by digint** [comment]
As you state correctly, this is mainly about security. As long as I have to tell people that btrbk needs root ssh logins (except if you use  [btrfs-progs-btrbk](https://github.com/digint/btrfs-progs-btrbk), #113, which requires quite some knowledge), I really tend to disallow the "long ropes", because especially people who use weird file names usually have no idea of security implications.

Assembling shell commands with escaped whitespaces is a real pain, and I guess I could never get this done correctly (honestly I did not try, but as soon as remote shells are involved you sometimes need to double-escape things, check behavior of different shells, and then implement this all again in ssh_filter_btrbk.sh).

Another main problem when dealing with special characters is that btrbk relies on external tools, which all do different escaping. Some use backslash, others use character codes, even others use utf8, or even choose the encoding dynamically or system dependent. Of course this can all be handled, but adds a lot of complexity and headache.

That said, I agree that it should still be possible to use such paths, eventually with an "I know what I'm doing / I don't care" option.

#### Regarding your proposed solution:
Life would be much easier if `btrfs sub list` wouldn't print ambiguous relative paths (which happens as soon as you put subvolumes into regular directories). So originally I used the "real-path" from `btrfs sub show` in order to assemble root / canonical paths, honoring symlinks. Newer versions of btrbk resolve the root path by looking at /proc/self/mounts, which is used as fallback, but has its own quirks (think of chroot for example).
This all works pretty well, I do have a canonical path to work with, and I guess I could handle whitespaces correctly.

#### Summary

  - yes, this is all doable
  - it adds complexity
  - it comes with security implications
  - it gives me headache

I guess I'll give it a try when I get some free time for this, especially accepting whitespaces should not be too complex.

**Comment by zatricky** [comment]
Thank you. I certainly appreciate the complexity. And thank you for this awesome tool. :)

**Comment by digint** [comment]
Implemented in btrbk-0.32.0, closing issue
